src/passes/call-utils.h

```diff
@@ -0,0 +1,153 @@
+//    (call $A)
+//    (call $B)
+//  )
+//
+// If we fail to find the expected pattern, or we decide it is not worth
+// optimizing it for some reason, we return nullptr.
+//
+// If this returns the new form, it will modify the function as necessary,
+// adding new locals etc., which later passes should optimize.
+//
+// |getCallInfo| is given one of the arms of the select and should return an
+// IndirectCallInfo that says what we know about it. We may know nothing, or
+// that it will trap, or that it will go to a known static target.
+template<typename T>
+inline Expression*
+convertToDirectCalls(T* curr,
+                     std::function<IndirectCallInfo(Expression*)> getCallInfo,
+                     Function& func,
+                     Module& wasm) {
+  auto* select = curr->target->template dynCast<Select>();
+  if (!select) {
+    return nullptr;
+  }
+
+  if (select->condition->type == Type::unreachable) {
+    // Leave this for DCE.
+    return nullptr;
+  }
+
+  // Check if we can find useful info for both arms: either known call targets,
+  // or traps.
+  // TODO: support more than 2 targets (with nested selects)
+  auto ifTrueCallInfo = getCallInfo(select->ifTrue);
+  auto ifFalseCallInfo = getCallInfo(select->ifFalse);
+  if (std::get_if<Unknown>(&ifTrueCallInfo) ||
+      std::get_if<Unknown>(&ifFalseCallInfo)) {
+    // We know nothing about at least one arm, so give up.
+    // TODO: Perhaps emitting a direct call for one arm is enough even if the
+    //       other remains indirect?
+    return nullptr;
+  }
+
+  auto& operands = curr->operands;
+
+  // We must use the operands twice, and also must move the condition to
+  // execute first; use locals for them all. While doing so, if we see
+  // any are unreachable, stop trying to optimize and leave this for DCE.
+  std::vector<Index> operandLocals;
+  for (auto* operand : operands) {
+    if (operand->type == Type::unreachable ||
+        !TypeUpdating::canHandleAsLocal(operand->type)) {
+      return nullptr;
+    }
+  }
+
+  Builder builder(wasm);
+  std::vector<Expression*> blockContents;
+
+  // None of the types are a problem, so we can proceed to add new vars as
+  // needed and perform this optimization.
+  for (auto* operand : operands) {
+    auto currLocal = builder.addVar(&func, operand->type);
+    operandLocals.push_back(currLocal);
+    blockContents.push_back(builder.makeLocalSet(currLocal, operand));
+  }
+
+  // Build the calls.
+  auto numOperands = operands.size();
+  auto getOperands = [&]() {
+    std::vector<Expression*> newOperands(numOperands);
+    for (Index i = 0; i < numOperands; i++) {
+      newOperands[i] =
+        builder.makeLocalGet(operandLocals[i], operands[i]->type);
+    }
+    return newOperands;
+  };
+
+  auto makeCall = [&](IndirectCallInfo info) -> Expression* {
+    if (std::get_if<Trap>(&info)) {
+      return builder.makeUnreachable();
+    } else {
+      return builder.makeCall(std::get<Known>(info).target,
```

Comment:
Did we solve the problem where `std::get` pulled in some library code that didn't exist on the Mac platforms we were targeting?

Reply:
I hope so :smile: A grep shows 3 places we use `std::get` before this PR adds another that makes it 4.